# Notebook to prep functions for report

* Grouped down to the DISTRICT and the CYCLE level

In [1]:
import intake
import numpy as np
import pandas as pd
import geopandas as gpd
from calitp import to_snakecase
from dla_utils import _dla_utils
from IPython.display import HTML, Markdown
from siuba import *
from shared_utils import geography_utils

import altair as alt

import _data_cleaning
import _report_utils

/opt/conda/lib/python3.10/site-packages/geopandas/_compat.py:123: UserWarning: The Shapely GEOS version (3.10.3-CAPI-1.16.1) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.
  warnings.warn(


In [2]:
import fiona

from ipyleaflet import Map, GeoJSON, projections, basemaps, GeoData, LayersControl, WidgetControl, GeoJSON, LegendControl
from ipywidgets import Text, HTML

In [3]:
GCS_FILE_PATH = 'gs://calitp-analytics-data/data-analyses/dla/atp/'


In [4]:
df_all = _report_utils.read_in_joined_data()

/home/jovyan/data-analyses/dla/atp/_report_utils.py:67: FutureWarning: In a future version, the Index constructor will not infer numeric dtypes when passed object-dtype sequences (matching Series behavior)


In [5]:
pd.set_option("display.max_columns",500)

In [6]:
## paramaters cell
district = 4

In [7]:
## parameters cell
cycle = 5

In [8]:
## subset df to just the district
#df = df_all>>filter(_.a2_ct_dist==district)

In [9]:
df = df_all.copy()

In [10]:
df = df>>filter(_.project_cycle==cycle)

In [11]:
# #check where everything is mapped
# df = df_all.copy()

In [12]:
display(HTML("<h2>Quick Stats</h2>"))

display(HTML(f"Out of {len(df)} Active Transportation Program Project Applications, "
            f"there are <strong>{(df>>filter(_.awarded=='Y')).project_app_id.nunique()} "
            f"projects</strong> that recieved funding over "
            f"{df.project_cycle.nunique()} cycles"))


HTML(value='<h2>Quick Stats</h2>')

HTML(value='Out of 450 Active Transportation Program Project Applications, there are <strong>49 projects</stro…

In [13]:
display(HTML(f"<h3> What were the application outcomes "
             f"for District {district} "
             f"in Cycle {cycle}?</h3>"))
display(HTML(_dla_utils.pretify_tables((df>>count(_.data_origin)))))

HTML(value='<h3> What were the application outcomes for District 4 in Cycle 5?</h3>')

HTML(value='<style type="text/css">\n#T_e9a1d th {\n  text-align: center;\n}\n#T_e9a1d_row0_col0, #T_e9a1d_row…

In [14]:
quick_view = df>>filter(_.awarded=="Y")>>select(_.data_origin, _.a1_imp_agcy_name, _.a2_info_proj_name, 
                                   _.a2_county, _.total_project_cost)

In [15]:
quick_view['total_project_cost'] = quick_view['total_project_cost'].map('$ {:0,.2f}'.format)

In [16]:
display(HTML("<h3> Funded Projects </h3>"))
display(HTML(_dla_utils.pretify_tables(quick_view)))

HTML(value='<h3> Funded Projects </h3>')

HTML(value='<style type="text/css">\n#T_9fcbb th {\n  text-align: center;\n}\n#T_9fcbb_row0_col0, #T_9fcbb_row…

In [17]:
#df>>group_by(_.awarded)>>count(_.a2_county)>>arrange(_.a2_county)

## Mapping

In [18]:
df_map= (df >>select(_.awarded, _.project_app_id, _.project_cycle, _.data_origin, _.geometry,
               _.a1_imp_agcy_city, _.a1_imp_agcy_name, _.a1_proj_partner_agcy, 
               _.assembly_district, _.congressional_district, _.senate_district,
              _.a2_county, _.a2_info_proj_descr, _.a2_info_proj_loc, _.a2_info_proj_name,
               _.a2_mop_uza_population, _.a2_mpo, _.a1_imp_agcy_street, _.a3_proj_type, 
               _.a3_proj_type, _['total_atp_$'], _.a2_proj_lat, _.a2_proj_long))

In [19]:
df_map = df_map>>filter(_.project_cycle==5)

In [20]:
#df_map>>filter(_.geometry.isnull())

In [21]:
## reapply geometry col for lat long
df_map = (geography_utils.create_point_geometry(df_map, longitude_col = 'a2_proj_long', latitude_col = 'a2_proj_lat'))

#### Flag for Incorrect Coordinates

In [22]:
## checking which is in/out of state bounds using

In [23]:
ca_bounds = gpd.read_parquet('gs://calitp-analytics-data/data-analyses/high_quality_transit_areas/ca_boundary.parquet')

In [24]:
ca_bounds

State                                           geometry
0    CA  MULTIPOLYGON (((-122.52997 37.81541, -122.5300...

In [25]:
#ca_bounds.explore()  

In [26]:
#gpd.overlay(ca_bounds, df_map3, how='difference')
# df_map3['in_ca'] = ""
# df_map3['in_ca'] = 
#df_map3.overlay(ca_bounds, how = "symmetric_difference")

In [27]:
(ca_bounds.contains(df_map))

/opt/conda/lib/python3.10/site-packages/geopandas/base.py:31: UserWarning: The indices of the two GeoSeries are different.


0      False
0       True
1      False
1      False
2      False
       ...  
443    False
444    False
445    False
446    False
447    False
Length: 450, dtype: bool

In [28]:
joined = (df_map.sjoin(ca_bounds, how='left'))

In [29]:
joined.State.value_counts()

CA    320
Name: State, dtype: int64

In [30]:
len(joined>>filter(_.State.isna()))

130

In [31]:
(joined>>filter(_.State.isna()))

awarded                                     project_app_id project_cycle  \
325       N                             6-Kingsburg, City of-1             5   
89        N                                6-Delano, City of-2             5   
271       N                                6-Fowler, City of-1             5   
378       N                                7-Avalon, City of-1             5   
382       N                          7-Baldwin Park, City of-1             5   
..      ...                                                ...           ...   
13        Y  11-San Diego Association of Governments (SANDA...          5.00   
29        Y                            4-Contra Costa County-2          5.00   
28        Y                               10-Mariposa County-2          5.00   
1         Y                        7-South El Monte, City of-1          5.00   
38        Y                               10-Tuolumne County-1          5.00   

     data_origin                            geometry a1_imp_agcy_city  \
325  Application          POINT (119.56000 36.51000)        Kingsburg   
89   Application          POINT (119.24710 35.76880)           Delano   
271  Application          POINT (119.68000 36.64000)           Fowler   
378  Application          POINT (118.33000 33.34000)           Avalon   
382  Application          POINT (117.96491 43.10680)     Baldwin Park   
..           ...                                 ...              ...   
13        Funded  POINT (1170624.00000 324511.00000)        San Diego   
29        Funded          POINT (121.94196 38.02416)         Martinez   
28        Funded          POINT (119.97080 37.48700)         Mariposa   
1         Funded          POINT (118.04670 34.05200)   South El Monte   
38        Funded          POINT (120.23264 37.58579)           Sonora   

                                  a1_imp_agcy_name a1_proj_partner_agcy  \
325                             Kingsburg, City of                  NaN   
89                                 Delano, City of                  NaN   
271                                Fowler, City of                  NaN   
378                                Avalon, City of                  NaN   
382                          Baldwin Park, City of                  NaN   
..                                             ...                  ...   
13   San Diego Association of Governments (SANDAG)    City of San Diego   
29                             Contra Costa County                  NaN   
28                                 Mariposa County                  NaN   
1                          South El Monte, City of                  NaN   
38                                 Tuolumne County                  NaN   

    assembly_district congressional_district senate_district     a2_county  \
325                31                     21              12        Fresno   
89                 32                     21              14          Kern   
271                31                     21              16        Fresno   
378                70                     47              26   Los Angeles   
382                48                     32              22   Los Angeles   
..                ...                    ...             ...           ...   
13             78, 80                 51, 53              40     San Diego   
29                 14                     11               7  Contra Costa   
28                  5                      4               8      Mariposa   
1                  57                     38              22   Los Angeles   
38                  8                      4               8      Tuolumne   

                                    a2_info_proj_descr  \
325  Roosevelt and Reagan Elementary School Pedestr...   
89   Construct 6,547 feet of new 4.5 ft. wide sidew...   
271  Fremont Elementary/Marshall Elementary/Fowler ...   
378  Conversion of the "Five-Corners" intersections...   
382  Install Traffic Control Devices, 

In [32]:
_report_utils.check_point_in_state(joined,
                        'State',
                        'CA')

awarded                   project_app_id project_cycle  data_origin  \
180       N            03-El Dorado County-1             5  Application   
99        N              6-Fresno, City of-2             5  Application   
211       N              6-Fresno, City of-3             5  Application   
275       N  6-Kern Council of Governments-1             5  Application   
415       N             6-Kern County - D6-1             5  Application   
..      ...                              ...           ...          ...   
42        Y         7-Los Angeles, City of-5          5.00       Funded   
41        Y               6-Huron, City of-1          5.00       Funded   
38        Y             10-Tuolumne County-1          5.00       Funded   
40        Y          7-Long Beach, City of-1          5.00       Funded   
35        Y               7-Ventura County-2          5.00       Funded   

                        geometry a1_imp_agcy_city  \
180  POINT (-120.82612 38.70348)      Placerville   
99   POINT (-119.80835 36.75038)           Fresno   
211  POINT (-119.77249 36.76316)           Fresno   
275  POINT (-119.01526 35.37227)      Bakersfield   
415  POINT (-118.89088 35.43151)      Bakersfield   
..                           ...              ...   
42   POINT (-118.45719 34.22689)      Los Angeles   
41   POINT (-120.10304 36.21027)            Huron   
38    POINT (120.23264 37.58579)           Sonora   
40   POINT (-118.19233 33.77071)       Long Beach   
35   POINT (-119.16290 34.23514)          Ventura   

                a1_imp_agcy_name a1_proj_partner_agcy assembly_district  \
180             El Dorado County                  NaN                 5   
99               Fresno, City of                  NaN            23, 31   
211              Fresno, City of                  NaN                31   
275  Kern Council of Governments                  NaN        32, 34, 36   
415             Kern County - D6                  NaN                34   
..                           ...                  ...               ...   
42          Los Angeles, City of                  NaN                46   
41                Huron, City of                  NaN                31   
38               Tuolumne County                  NaN                 8   
40           Long Beach, City of                  NaN                70   
35                Ventura County                  NaN                44   

    congressional_district          senate_district    a2_county  \
180                      4                        1    El Dorado   
99                      16  Needs Manual Assistance       Fresno   
211                     16  Needs Manual Assistance       Fresno   
275                 21, 23                   14, 16         Kern   
415                     23                       16         Kern   
..                     ...                      ...          ...   
42                      29                       18  Los Angeles   
41                      21                       12       Fresno   
38                       4                        8     Tuolumne   
40                      47                       33  Los Angeles   
35                      26                       19      Ventura   

                                    a2_info_proj_descr  \
180  Construct a Class I multi-use grade-separated ...   
99   Palm and Belmont: 3.5 miles of Class IV with C...   
211  Install signal/scramble at Chestnut & Weldon, ...   
275  Identified disadvantaged communities in Kern C...   
415  Rehabilitate & resurface the Kern River Parkwa...   
..                                                 ...   
42   Transform one of the City’s most traffic-stres...   
41   The project will add Class II, III and IV bike...   
38   Project includes approximately 4,200 linear fe...   
40   Transform Downtown Long Beach into a pedestria...   
35   Install sidewalks, curbs, gutters, water-reten...   

                                      a2_info_proj_

In [33]:
joined>>filter(_.a2_proj_lat>300)

awarded                                     project_app_id project_cycle  \
283       N                              7-Pasadena, City of-1             5   
260       N                              7-Pasadena, City of-2             5   
13        Y  11-San Diego Association of Governments (SANDA...          5.00   

     data_origin                             geometry a1_imp_agcy_city  \
283  Application  POINT (1883071.21000 6516102.76300)         Pasadena   
260  Application  POINT (1883071.21000 6516102.76300)         Pasadena   
13        Funded   POINT (1170624.00000 324511.00000)        San Diego   

                                  a1_imp_agcy_name a1_proj_partner_agcy  \
283                              Pasadena, City of                  NaN   
260                              Pasadena, City of                  NaN   
13   San Diego Association of Governments (SANDAG)    City of San Diego   

    assembly_district congressional_district senate_district    a2_county  \
283                41                     27              25  Los Angeles   
260                41                     27              25  Los Angeles   
13             78, 80                 51, 53              40    San Diego   

                                    a2_info_proj_descr  \
283  The goal of the project is to develop ATP for ...   
260  Replacement of existing standard crosswalks wi...   
13   Buffered bike lanes; traffic calming and ped a...   

                                      a2_info_proj_loc  \
283  Northwest Pasadena: 210 Freeway to the South, ...   
260  Northwest Pasadena: 210 Freeway to the south, ...   
13   The Orange Family Friendly Street Project is 2...   

                                     a2_info_proj_name  \
283  Northwest Pasadena Active Transportation Plan ...   
260  Northwest Pasadena Continental Crosswalk Imple...   
13               Orange Family Friendly Street Project   

                                 a2_mop_uza_population  a2_mpo  \
283  Project is located within one of the ten large...    SCAG   
260  Project is located within one of the ten large...    SCAG   
13   Project is located within one of the ten large...  SANDAG   

            a1_imp_agcy_street            a3_proj_type  total_atp_$  \
283  221 East Walnut Suite 210                    Plan          NaN   
260     221 East Walnut Street  Infrastructure - Small          NaN   
13     401 B Street, Suite 800                     NaN 4,317,000.00   

     a2_proj_lat  a2_proj_long  index_right State         point_check  
283 6,516,102.76  1,883,071.21          NaN  None  Point Not In State  
260 6,516,102.76  1,883,071.21          NaN  None  Point Not In State  
13    324,511.00  1,170,624.00          NaN  None  Point Not In State

In [34]:
joined2 = joined>>filter(_.a2_proj_lat<300)

In [35]:
## map shows multiple points out of CA
#joined2.explore("point_check", cmap="tab20b")  

In [36]:
joined2.point_check.value_counts()

Point In State        320
Point Not In State    127
Name: point_check, dtype: int64

In [37]:
(joined2>>filter(_.point_check=='Point Not In State')
 >>select(_.data_origin, _.project_app_id,_.a1_imp_agcy_city, _.geometry, _.a2_proj_lat, _.a2_proj_long)
 >>arrange(_.a2_proj_long)).iloc[1:50]

data_origin                                     project_app_id  \
28   Application                     1-Round Valley Indians Tribe-1   
214  Application                         5-San Luis Obispo County-2   
351  Application                              7-Cerritos, City of-2   
23        Funded                        11-National City, City of-3   
186  Application                               8-Needles, City of-1   
402  Application                        11-National City, City of-1   
277  Application                          11-Chula Vista, City of-1   
43        Funded                       11-Imperial Beach, City of-1   
247  Application                               8-Riverside County-3   
241  Application                            11-San Diego, City of-4   
119  Application  11-San Diego Association of Governments (SANDA...   
279  Application                         11-Solana Beach, City of-1   
155  Application                            11-Encinitas, City of-1   
284  Application                            11-Oceanside, City of-3   
160  Application                                 12-Orange County-3   
384  Application                         12-Laguna Hills, City of-1   
11   Application                             8-Montclair, City of-1   
343  Application                               12-Irvine, City of-1   
85   Application                           12-Santa Ana, City of-21   
20   Application                           12-Santa Ana, City of-15   
1    Application                            12-Santa Ana, City of-4   
44   Application                            12-Santa Ana, City of-8   
4    Application                           12-Santa Ana, City of-14   
94   Application                           12-Santa Ana, City of-16   
15   Application                            12-Santa Ana, City of-2   
10   Application                            12-Santa Ana, City of-7   
22   Application                           12-Santa Ana, City of-17   
14   Application                            12-Santa Ana, City of-9   
21   Application                            12-Santa Ana, City of-1   
92   Application                           12-Santa Ana, City of-18   
3    Application                           12-Santa Ana, City of-13   
87   Application                            12-Santa Ana, City of-3   
68   Application                            12-Santa Ana, City of-6   
13   Application                            12-Santa Ana, City of-5   
95   Application                           12-Santa Ana, City of-11   
74   Application                           12-Santa Ana, City of-12   
79   Application                           12-Santa Ana, City of-10   
29   Application                           12-Santa Ana, City of-20   
93   Application                           12-Santa Ana, City of-19   
36   Application                           7-West Covina, City of-1   
382  Application                          7-Baldwin Park, City of-1   
387  Application        7-California Department of Transportation-1   
1         Funded                        7-South El Monte, City of-1   
316  Application                           7-Culver City, City of-2   
368  Application                              7-Alhambra, City of-1   
37   Application                              7-Commerce, City of-1   
179  Application                            7-Long Beach, City of-3   
445  Application                            7-Long Beach, City of-4   
396  Application                              7-Glendale, City of-1   

          a1_imp_agcy_city                     geometry  a2_proj_lat  \
28                  Covelo  POINT (-125.26222 39.79472)        39.79   
214        San Luis Obispo  POINT (-120.87421 31.41508)        31.42   
351               Cerritos  POINT (-118.04340 33.50450)        33.50   
23           National City    POINT (32.65651 32.65651)        32.66   
186                Needles   POINT (114.36110 34.50040)        34.50   
402          National 

* most longitudes are positive, when they should be negative, resulting in the longitudes being in east Asia

#### QUICK FIX for wrong longs
* funded data lost mapping data from geometry column

In [38]:
df_map_correct = (joined2>>filter(_.a2_proj_long<0))

In [39]:
df_map_incorrect = (joined2>>filter(_.a2_proj_long>0))

In [40]:
df_map_incorrect['a2_proj_long'] = df_map_incorrect['a2_proj_long'] * (-1)

/opt/conda/lib/python3.10/site-packages/geopandas/geodataframe.py:1443: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [41]:
df_map_corrected = (geography_utils.create_point_geometry(df_map_incorrect, longitude_col = 'a2_proj_long', latitude_col = 'a2_proj_lat'))

In [42]:
df_map2 = (pd.concat([df_map_corrected, df_map_correct]))

In [43]:
df_map2 = df_map2.drop(columns=['index_right','State','point_check']) 

In [44]:
joined3 = (df_map2.sjoin(ca_bounds, how='left'))

In [45]:
joined3 = _report_utils.check_point_in_state(joined3,
                        'State',
                        'CA')

In [46]:
joined3.point_check.value_counts()

Point In State        437
Point Not In State     10
Name: point_check, dtype: int64

* after changing the positive longitudes to negative, the number of points <strong>"Not In State"</strong> went from 127 to 10

In [47]:
#check again 
joined3.explore("point_check", cmap="tab20b")  

In [48]:
# still have some weird locations but better than before

In [49]:
## going back to just district

In [50]:
need_help = (joined3>>filter(_.point_check=='Point Not In State') 
 >>select(_.data_origin, _.project_app_id, _.a1_imp_agcy_city, _.geometry, _.a2_proj_lat, _.a2_proj_long)
 >>arrange(_.a2_proj_long)).reset_index()

In [51]:
need_help

index  data_origin                               project_app_id  \
0     73  Application  9-California Department of Transportation-6   
1    232  Application                   8-Jurupa Valley, City of-1   
2     28  Application               1-Round Valley Indians Tribe-1   
3    214  Application                   5-San Luis Obispo County-2   
4    316  Application                     7-Culver City, City of-2   
5    351  Application                        7-Cerritos, City of-2   
6    382  Application                    7-Baldwin Park, City of-1   
7    155  Application                      11-Encinitas, City of-1   
8    186  Application                         8-Needles, City of-1   
9     23       Funded                  11-National City, City of-3   

  a1_imp_agcy_city                     geometry  a2_proj_lat  a2_proj_long  
0           Bishop  POINT (-188.42980 37.36138)        37.36       -188.43  
1    Jurupa Valley  POINT (-177.42278 33.99897)        34.00       -177.42  
2           Covelo  POINT (-125.26222 39.79472)        39.79       -125.26  
3  San Luis Obispo  POINT (-120.87421 31.41508)        31.42       -120.87  
4      Culver City  POINT (-118.10346 33.47235)        33.47       -118.10  
5         Cerritos  POINT (-118.04340 33.50450)        33.50       -118.04  
6     Baldwin Park  POINT (-117.96491 43.10680)        43.11       -117.96  
7        Encinitas  POINT (-117.28330 32.02250)        32.02       -117.28  
8          Needles  POINT (-114.36110 34.50040)        34.50       -114.36  
9    National City   POINT (-32.65651 32.65651)        32.66        -32.66

In [52]:
### adding in census geojson
### following code does not work

#shape = gpd.read_file('gs://calitp-analytics-data/data-analyses/dla/atp/CA_Places_TIGER2016.shp')
#census_bound = gpd.read_file('gs://calitp-analytics-data/data-analyses/census_boundaries/boundary_tract_2018.geojson')

In [53]:
places = "https://data.ca.gov/dataset/e212e397-1277-4df3-8c22-40721b095f33/resource/436fc714-831c-4070-b44b-b06dcde6bf18/download/ca-places-boundaries.zip"
places_ca = gpd.read_file(places)

In [54]:
places_ca = places_ca>>select(_.NAME, _.NAMELSAD, _.INTPTLAT, _.INTPTLON, _.geometry)

In [55]:
places_ca['geometry2'] = places_ca['geometry'].centroid

In [56]:
places_ca.sample(4)

NAME          NAMELSAD     INTPTLAT      INTPTLON  \
860  Graniteville  Graniteville CDP  +39.4443656  -120.7360303   
90        Portola      Portola city  +39.8207990  -120.4742930   
455    Fort Bragg   Fort Bragg city  +39.4410694  -123.8036201   
494     Alleghany     Alleghany CDP  +39.4666599  -120.8411238   

                                              geometry  \
860  POLYGON ((-13442540.754 4783484.331, -13442535...   
90   MULTIPOLYGON (((-13410323.892 4839537.975, -13...   
455  MULTIPOLYGON (((-13779179.962 4785501.483, -13...   
494  POLYGON ((-13452813.096 4788380.765, -13452806...   

                             geometry2  
860  POINT (-13440274.641 4785525.509)  
90   POINT (-13411022.309 4840421.488)  
455  POINT (-13781733.462 4785109.405)  
494  POINT (-13451973.851 4788739.611)

In [57]:
##check points to make sure 
#places_ca.explore("NAME")  

In [58]:
# places_geom = dict(zip(places_ca['NAME'], 
#                           places_ca['geometry2']))

In [59]:
#need_help['geometry3'] = df['a1_imp_agcy_city'].map(places_geom)

In [60]:
places_ca= places_ca>>select(_.NAME, _.NAMELSAD, _.INTPTLAT, _.INTPTLON, _.geometry2)

In [61]:
places_ca.sample()

NAME           NAMELSAD     INTPTLAT      INTPTLON  \
1346  Sleepy Hollow  Sleepy Hollow CDP  +38.0120356  -122.5876053   

                              geometry2  
1346  POINT (-13646391.710 4581127.146)

In [62]:
need_help2 = ((pd.merge(need_help, places_ca,
           how = 'left',
           left_on='a1_imp_agcy_city',
           right_on='NAME'))>>select(_.data_origin, _.project_app_id, _.a1_imp_agcy_city, _.NAME,
                                     _.NAMELSAD,_.INTPTLAT,_.INTPTLON))

In [63]:
need_help2

data_origin                               project_app_id a1_imp_agcy_city  \
0  Application  9-California Department of Transportation-6           Bishop   
1  Application                   8-Jurupa Valley, City of-1    Jurupa Valley   
2  Application               1-Round Valley Indians Tribe-1           Covelo   
3  Application                   5-San Luis Obispo County-2  San Luis Obispo   
4  Application                     7-Culver City, City of-2      Culver City   
5  Application                        7-Cerritos, City of-2         Cerritos   
6  Application                    7-Baldwin Park, City of-1     Baldwin Park   
7  Application                      11-Encinitas, City of-1        Encinitas   
8  Application                         8-Needles, City of-1          Needles   
9       Funded                  11-National City, City of-3    National City   

              NAME              NAMELSAD     INTPTLAT      INTPTLON  
0           Bishop           Bishop city  +37.3663813  -118.3958082  
1    Jurupa Valley    Jurupa Valley city  +34.0025907  -117.4676122  
2           Covelo            Covelo CDP  +39.8001872  -123.2526012  
3  San Luis Obispo  San Luis Obispo city  +35.2639540  -120.6611260  
4      Culver City      Culver City city  +34.0058204  -118.3967807  
5         Cerritos         Cerritos city  +33.8677431  -118.0694719  
6     Baldwin Park     Baldwin Park city  +34.0828245  -117.9712858  
7        Encinitas        Encinitas city  +33.0506323  -117.2636163  
8          Needles          Needles city  +34.8135697  -114.6253021  
9    National City    National City city  +32.6658617  -117.0973610

In [64]:
need_help2 = (geography_utils.create_point_geometry(need_help2,
                                                    longitude_col = 'INTPTLON',
                                                    latitude_col = 'INTPTLAT'))

In [65]:
need_help2

data_origin                               project_app_id a1_imp_agcy_city  \
0  Application  9-California Department of Transportation-6           Bishop   
1  Application                   8-Jurupa Valley, City of-1    Jurupa Valley   
2  Application               1-Round Valley Indians Tribe-1           Covelo   
3  Application                   5-San Luis Obispo County-2  San Luis Obispo   
4  Application                     7-Culver City, City of-2      Culver City   
5  Application                        7-Cerritos, City of-2         Cerritos   
6  Application                    7-Baldwin Park, City of-1     Baldwin Park   
7  Application                      11-Encinitas, City of-1        Encinitas   
8  Application                         8-Needles, City of-1          Needles   
9       Funded                  11-National City, City of-3    National City   

              NAME              NAMELSAD     INTPTLAT      INTPTLON  \
0           Bishop           Bishop city  +37.3663813  -118.3958082   
1    Jurupa Valley    Jurupa Valley city  +34.0025907  -117.4676122   
2           Covelo            Covelo CDP  +39.8001872  -123.2526012   
3  San Luis Obispo  San Luis Obispo city  +35.2639540  -120.6611260   
4      Culver City      Culver City city  +34.0058204  -118.3967807   
5         Cerritos         Cerritos city  +33.8677431  -118.0694719   
6     Baldwin Park     Baldwin Park city  +34.0828245  -117.9712858   
7        Encinitas        Encinitas city  +33.0506323  -117.2636163   
8          Needles          Needles city  +34.8135697  -114.6253021   
9    National City    National City city  +32.6658617  -117.0973610   

                      geometry  
0  POINT (-118.39581 37.36638)  
1  POINT (-117.46761 34.00259)  
2  POINT (-123.25260 39.80019)  
3  POINT (-120.66113 35.26395)  
4  POINT (-118.39678 34.00582)  
5  POINT (-118.06947 33.86774)  
6  POINT (-117.97129 34.08282)  
7  POINT (-117.26362 33.05063)  
8  POINT (-114.62530 34.81357)  
9  POINT (-117.09736 32.66586)

In [66]:
## mapping to check if the points match. it does.
need_help2.explore("data_origin")

### Adding Flag for corrected geometries

In [67]:
#### Check what the geometry looks like for the points that do not fall within CA State Bounds

In [68]:
joined>>filter(_.point_check=='Point Not In State')>>select(_.awarded, _.geometry, _.a1_imp_agcy_city,
                                                            _.a1_imp_agcy_name, _.a2_county, _.a2_proj_lat,
                                                            _.a2_proj_long, _.index_right, _.State, _.point_check)>>arrange(_.a2_proj_lat)


awarded                             geometry a1_imp_agcy_city  \
214       N          POINT (-120.87421 31.41508)  San Luis Obispo   
155       N           POINT (117.28330 32.02250)        Encinitas   
43        Y           POINT (117.11441 32.57646)   Imperial Beach   
277       N           POINT (117.09559 32.63669)      Chula Vista   
23        Y            POINT (32.65651 32.65651)    National City   
..      ...                                  ...              ...   
317       N           POINT (120.17330 41.52910)          Alturas   
382       N           POINT (117.96491 43.10680)     Baldwin Park   
13        Y   POINT (1170624.00000 324511.00000)        San Diego   
283       N  POINT (1883071.21000 6516102.76300)         Pasadena   
260       N  POINT (1883071.21000 6516102.76300)         Pasadena   

                                  a1_imp_agcy_name        a2_county  \
214                         San Luis Obispo County  San Luis Obispo   
155                             Encinitas, City of        San Diego   
43                         Imperial Beach, City of        San Diego   
277                           Chula Vista, City of        San Diego   
23                          National City, City of        San Diego   
..                                             ...              ...   
317                                   Modoc County            Modoc   
382                          Baldwin Park, City of      Los Angeles   
13   San Diego Association of Governments (SANDAG)        San Diego   
283                              Pasadena, City of      Los Angeles   
260                              Pasadena, City of      Los Angeles   

     a2_proj_lat  a2_proj_long  index_right State         point_check  
214        31.42       -120.87          NaN  None  Point Not In State  
155        32.02        117.28          NaN  None  Point Not In State  
43         32.58        117.11          NaN  None  Point Not In State  
277        32.64        117.10          NaN  None  Point Not In State  
23         32.66         32.66          NaN  None  Point Not In State  
..           ...           ...          ...   ...                 ...  
317        41.53        120.17          NaN  None  Point Not In State  
382        43.11        117.96          NaN  None  Point Not In State  
13    324,511.00  1,170,624.00          NaN  None  Point Not In State  
283 6,516,102.76  1,883,071.21          NaN  None  Point Not In State  
260 6,516,102.76  1,883,071.21          NaN  None  Point Not In State  

[130 rows x 10 columns]

## Metrics

In [69]:
unique_agencies = df>>group_by(_.project_cycle, _.a2_county, _.data_origin)>>summarize(n_unique_agency= _.a1_imp_agcy_name.nunique())

In [70]:
unique_agencies =(spread(unique_agencies, "data_origin", "n_unique_agency"))

In [71]:
unique_agencies = unique_agencies.rename(columns={"a2_county":"county_name"})

In [72]:
unique_agencies['Application'].fillna(0, inplace=True)
unique_agencies['Funded'].fillna(0, inplace=True)

In [73]:
unique_agencies['Application'] = unique_agencies['Application'].astype('int32')
unique_agencies['Funded'] = unique_agencies['Funded'].astype('int32')

In [74]:
display(HTML("<h3>Number of Unique Agencies By County</h3>"))
display(HTML(_dla_utils.pretify_tables(unique_agencies)))


HTML(value='<h3>Number of Unique Agencies By County</h3>')

HTML(value='<style type="text/css">\n#T_f54da th {\n  text-align: center;\n}\n#T_f54da_row0_col0, #T_f54da_row…

### Success Rates

In [75]:
df = (_report_utils.reorder_namecol(df,
                    og_name_col= 'a1_imp_agcy_name',
                    new_name_col= 'imp_agency_name_new', 
                    split_on = ", ",
                   order_on ='pt2_pt1'))

/home/jovyan/data-analyses/dla/atp/_report_utils.py:202: FutureWarning: In a future version of pandas all arguments of StringMethods.split except for the argument 'pat' will be keyword-only.


In [76]:
successes = (df>>group_by(_.awarded)>>count(_.imp_agency_name_new))>>spread("awarded", "n")>>arrange(-_.Y)

In [77]:
successes['total'] = (successes['N'] + successes['Y'])

In [78]:
successes['success_rate'] = (successes['Y']/successes['total'])

In [79]:
successes = successes.rename(columns={"imp_agency_name_new":"Implementing Agency", "N":"Projects Not Funded",
                                    "Y":"Funded Projects", "total":"Total Applications"})

In [80]:
successes_top = successes>>filter(_.success_rate>0)

In [81]:
successes_top['success_rate'] = successes_top['success_rate'].transform(lambda x: '{:,.2%}'.format(x))

/tmp/ipykernel_1969/3746543993.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [82]:
display(HTML("<h3>Application Success Rate</h3>"))

display(HTML(f"There are <strong>{len(successes>>filter(_.success_rate==0))}</strong> "
            f"implenting agencies with <strong> zero </strong>"
            f"successful applications."))

display(HTML(f"There are <strong>{len(successes>>filter(_.success_rate!=0))}</strong> "
            f"implenting agencies with <strong> one of more </strong>"
            f"successful applications."))
    
display(HTML("</br><h4> Success Rates for Agencies with Successful Applications </h4>"))
display(HTML(_dla_utils.pretify_tables(successes_top>>select(_['Implementing Agency'], _['Total Applications'], _.success_rate))))

HTML(value='<h3>Application Success Rate</h3>')

HTML(value='There are <strong>209</strong> implenting agencies with <strong> zero </strong>successful applicat…

HTML(value='There are <strong>42</strong> implenting agencies with <strong> one of more </strong>successful ap…

HTML(value='</br><h4> Success Rates for Agencies with Successful Applications </h4>')

HTML(value='<style type="text/css">\n#T_eaeb3 th {\n  text-align: center;\n}\n#T_eaeb3_row0_col0, #T_eaeb3_row…